src/tools/ceph_objectstore_tool.cc

```diff
@@ -680,55 +688,97 @@ int do_trim_pg_log(ObjectStore *store, const coll_t &coll,
-	cerr << "Error reading pg log entry: " << e.what() << std::endl;
-      }
-      if (debug) {
-	cerr << "read entry " << e << std::endl;
+
+      if (p->key().substr(0, 4) == string("dup_")) {
+        pg_log_dup_t dup;
+	try {
+	  dup.decode(bp);
+	  latest_dups.push_back(dup);
+	  if (latest_dups.size() > tracked_dups) {
+	    pg_log.dups.push_back(latest_dups.front());
+	    latest_dups.pop_front();
+	  }
+	  if (debug) {
+            cerr << "read entry " << dup << std::endl;
+          }
+	} catch (buffer::error& e) {
+	  cerr << "error reading log dup: " << p->key() << std::endl;
+	  return -EFAULT;
+	}
+      } else {
+        pg_log_entry_t e;
+        try {
+	  e.decode_with_checksum(bp);
+	  if (e.version > trim_to) {
+	    continue;
+	  }
+	  pg_log.log.push_back(e);
+	  if (debug) {
+            cerr << "read entry " << e << std::endl;
+          }
+	} catch (const buffer::error &e) {
+          cerr << "Error reading pg log entry: " << e.what() << std::endl;
+        }
       }
-      if (e.version.version > trim_to) {
-	done = true;
-	break;
+      // We have enough pg logs entries to trim
+      if ((pg_log.log.size() + pg_log.dups.size()) >= trim_at_once) {
+        break;
       }
-      keys_to_trim.insert(p->key());
-      new_tail = e.version;
-      if (keys_to_trim.size() >= trim_at_once)
-	break;
     }
-
     if (!p->valid())
       done = true;
+
+    pg_log.index();
+
     } // deconstruct ObjectMapIterator
 
-    // delete the keys
-    if (!dry_run && !keys_to_trim.empty()) {
-      cout << "Removing keys " << *keys_to_trim.begin() << " - " << *keys_to_trim.rbegin() << std::endl;
-      ObjectStore::Transaction t;
-      t.omap_rmkeys(coll, oid, keys_to_trim);
-      store->queue_transaction(ch, std::move(t));
-      ch->flush();
+    if (!dry_run && ((pg_log.log.size() > 0) || (pg_log.dups.size() > 0))) {
+      std::set<std::string> trimmed;
+      std::set<std::string> trimmed_dups;
+      eversion_t write_from_dups = eversion_t::max();
+
+      pg_log.trim(g_ceph_context, trim_to, &trimmed, &trimmed_dups, &write_from_dups);
+      // Is there any trimming to do?
+      if (!trimmed.empty() || !trimmed_dups.empty()) {
+        new_tail = pg_log.log.back().version;
+
+        ObjectStore::Transaction t;
+        if (!trimmed.empty()) {
+	  cerr << "Removing keys " << *trimmed.begin() << " - " << *trimmed.rbegin() << std::endl;
+          t.omap_rmkeys(coll, oid, trimmed);
+	}
+	if (!trimmed_dups.empty()) {
+          cerr << "Removing Dups " << *trimmed_dups.begin() << " - " << *trimmed_dups.rbegin() << std::endl;
+          t.omap_rmkeys(coll, oid, trimmed_dups);
+	}
+        store->queue_transaction(ch, std::move(t));
+        ch->flush();
+	pg_log.log.clear();
+	pg_log.dups.clear();
+      }
     }
   }
 
   // update pg info with new tail
   if (!dry_run && new_tail !=  eversion_t()) {
-    info.log_tail = new_tail;
+    if (latest_dups.size() > 0) {
+      info.log_tail = latest_dups.back().version;
```

Comment:
This looks suspicious to me. Isn't possible a situation when `latest_dups.size() > 0` but info.log_tail should be set to `new_tail`?

Reply:
need to set it always to new_tail. nothing to do with dups. fixed